mvc/models/model-binding.rst

```diff
@@ -1,8 +1,113 @@
+- ``[FromServices]``: Only supported on action parameters. 
+- ``[FromBody]``: Instructs the model binder to use the JSON media type formatter rather than regular model binding.
+- ``[ModelBinder]``: You reject the current model binder and substitute it with your own.
+
+A good example of binding attribute usage in the wild is when you see the ``[FromBody]`` attribute in action methods of API Controllers. While there are only a few differences between standard MVC and API Controllers, one such difference is that API Controllers attempt to retrieve data from the request URI in PUT and POST messages, rather than from the body by default. This means that when a page sends data via an Ajax call, Web API actions still look in the header, but applying the ``[FromBody]`` attribute will direct the model binder to the correct data source, as shown in the snippet below:
+
+.. literalinclude:: model-binding/sample/src/MVCMovie/Controllers/MoviesAPIController.cs
+   :language: c#
+   :lines: 47-49
+
+Binding attributes are very helpful tools when you need to nudge the default behaviors of model binders.
+
+Binding formatted data from the request body
+--------------------------------------------
+Formatting data is an important behavior of controllers, and it influences model binding behavior as well. A formatter is an object that accepts data and serializes or deserializes it into some shape. JSON is the most popular format, as it's the smallest and most lightweight way to structure textual data. It's also the default format for API controllers. Below is a list of several built-in formatters:  
+
+Input formatters:
+
+- ``JsonInputFormatter``: Based on JSON.NET, this object deserializes the requested resource from JSON into an object. 
+- ``JsonPatchInputFormatter``: Assists the JsonInputFormatter.
+
+Output formatters:
+
+- ``JsonOutputFormatter``: This is the counterpart to the JsonInputFormatter which serializes objects to JSON. 
+- ``HttpNoContentOutputFormatter``: This formatter handles responses with an HTTP status code 204, no content. 
+- ``StringOutputFormatter``: This formatter handles basic string responses.
+- ``StreamOutputFormatter``: This formatter handles outgoing streams.
+
+ASP.NET selects input formatters based on the Content-Type header and the type of the parameter, unless there is an attribute specifying otherwise. If you'd like to use XML or another format you must configure it in the Startup.cs file. You may have to install a NuGet package for the formatter you want. To see the formatters at runtime, set a breakpoint at the beginning of an action method and investigate the controller's ``ModelBindingContext`` object when in debug mode. You can find ``this``, which refers to the controller, in a debug window, as shown here:
+
+.. image:: model-binding/_static/formatters.png 
+
+Resources
+---------
+Below are links to the source code for model binding and formatters in ASP.NET MVC:
```

Comment:
We should avoid links to produce source because they are not stable. A link to the routing and response formatting would be good.